File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        length = 0
        temp = head
        while temp:
            temp = temp.next
            length += 1
        
        part_size = length // k
        extra_nodes = length % k
        
        result = []
        temp = head
        
        for i in range(k):
            if not temp:
                result.append(None)
                continue
            
            current_part_head = temp
            current_part_size = part_size + (1 if extra_nodes > 0 else 0)
            extra_nodes -= 1
            
            for _ in range(current_part_size - 1):
                temp = temp.next
            
            next_part_head = temp.next
            temp.next = None
            temp = next_part_head
            
            result.append(current_part_head)
        
        return result

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int len = 0;
        ListNode *temp = head;
        while (temp) {
            temp = temp->next;
            len++;
        }

        int partSize = len / k;
        int extraNodes = len % k;

        vector<ListNode*> ans(k, nullptr);
        temp = head;

        for (int i = 0; i < k; ++i) {
            if (!temp) break;
            ans[i] = temp;

            int currentPartSize = partSize + (extraNodes > 0 ? 1 : 0);
            extraNodes--;

            for (int j = 1; j < currentPartSize; ++j) {
                temp = temp->next;
            }

            ListNode* nextPart = temp->next;
            temp->next = nullptr;
            temp = nextPart;
        }

        return ans;
    }
};